src/translator/service.cpp

```diff
@@ -61,17 +61,26 @@ AsyncService::AsyncService(const AsyncService::Config &config)
   }
 }
 
-AsyncService::~AsyncService() {
+void AsyncService::terminate() {
+  safeBatchingPool_.clear();
+  join();
+}
+
+void AsyncService::join() {
   safeBatchingPool_.shutdown();
   for (std::thread &worker : workers_) {
     assert(worker.joinable());
     worker.join();
   }
+  workers_.clear();
 }
 
+AsyncService::~AsyncService() { join(); }
```

Comment:
This feels wrong. Join() and ~AsyncService() both do the same (finishing the translation and destroying all workers). Maybe at least `join()` should be private?

Reply:
I made it public so you could more easily write something that executes after all requests are processed without relying on the scope of the AsyncService object:
```c++
int main() {
  AsynService service;
  std::cerr << "Service started\n";

  std::thread reader([&]() {
    // reads lines and submits them as requests to service until stdin is empty
  });
  reader.join();
  std::cerr << "All requests submitted\n";

  service.join();
  std::cerr << "All requests handled\n";
  return 0;
}
```

Reply:
I guess... I prefer explicitly calling `delete service` at that point. As far as I can tell, once you call join, you can't "resurrect" the service, right?

Reply:
You can use delete if it would be heap-allocated. But if it is stack-allocated the only ways I know are more parentheses or wrapping it in a function to force it to go out of scope.

Reply:
oh, yeah. That. True.